File: src/models/TraversalOrderResult.ts
import type { Element } from "./Element";

/**
 * Result from getTraversalOrder WebSocket command containing ordered accessibility-focusable elements
 */
export interface TraversalOrderResult {
  /**
   * Ordered list of accessibility-focusable elements in TalkBack traversal order
   * (depth-first, left-to-right traversal)
   */
  elements: Element[];

  /**
   * Index of the currently focused element in the elements array.
   * null if no element is focused.
   */
  focusedIndex: number | null;

  /**
   * Total number of focusable elements found
   */
  totalCount: number;

  /**
   * Total time taken to extract traversal order in milliseconds
   */
  totalTimeMs: number;

  /**
   * Optional request ID for correlating responses
   */
  requestId?: string;

  /**
   * Error message if the operation failed
   */
  error?: string;
}

File: android/control-proxy/src/main/kotlin/dev/jasonpearson/automobile/ctrlproxy/models/TraversalOrderResult.kt
package dev.jasonpearson.automobile.ctrlproxy.models

import kotlinx.serialization.Serializable

/**
 * Result containing ordered accessibility-focusable elements in TalkBack traversal order.
 *
 * @property elements List of elements in traversal order (depth-first, left-to-right)
 * @property focusedIndex Index of currently focused element in the list, or null if no element has
 *   focus
 * @property totalCount Total number of focusable elements found
 */
@Serializable
data class TraversalOrderResult(
    val elements: List<UIElementInfo>,
    val focusedIndex: Int? = null,
    val totalCount: Int = elements.size,
)
